htsget/htsget.go

```diff
@@ -0,0 +1,238 @@
+				AcceptEncoding string `json:"accept-encoding"`
+				Authorization  string `json:"authorization"`
+			} `json:"headers,omitempty"`
+		} `json:"urls"`
+	} `json:"htsget"`
+}
+
+// Htsget function downloads the files included in the urls_list.txt file.
+// The argument can be a local file or a url to an S3 folder
+func Htsget(args []string) error {
+
+	// Call ParseArgs to take care of all the flag parsing
+	err := helpers.ParseArgs(args, Args)
+	if err != nil {
+		return fmt.Errorf("failed parsing arguments, reason: %v", err)
+	}
+
+	var currentPath string
+	currentPath, err = os.Getwd()
+	if err != nil {
+		return fmt.Errorf("failed to get current path, reason: %v", err)
+	}
+
+	if *datasetID == "" || *fileName == "" || *htsgetHost == "" || *publicKeyFile == "" {
+		return fmt.Errorf("missing required arguments, dataset, filename, htsgethost and key are required")
+	}
+
+	config, err := helpers.GetAuth(*configPath)
+	if err != nil {
+		return err
+	}
+
+	// read public key from file
+	publickey, err := os.ReadFile(currentPath + "/" + *publicKeyFile)
+	if err != nil {
+		return fmt.Errorf("failed to read public key, reason: %v", err)
+	}
+	base64publickey := base64.StdEncoding.EncodeToString(publickey)
+
+	// TODO: Add cases for different type of files
+	// i.e. bam files require the /reads/, replace for vcf
+	url := *htsgetHost + "/reads/" + *datasetID + "/" + *fileName
+	if *referenceName != "" {
+		url = url + "?referenceName=" + *referenceName
+	}
+
+	method := "GET"
+
+	client := &http.Client{}
+	req, err := http.NewRequest(method, url, nil)
+	if err != nil {
+		return fmt.Errorf("failed to make request, reason: %v", err)
+	}
+
+	req.Header.Add("Authorization", "Bearer "+config.AccessToken)
+	req.Header.Add("client-public-key", base64publickey)
+
+	res, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to do the request, reason: %v", err)
+	}
+	if res.StatusCode != 200 {
+		return fmt.Errorf("failed to get the file, status code: %v", res.StatusCode)
+	}
+	defer res.Body.Close()
+
+	body, err := io.ReadAll(res.Body)
+	if err != nil {
+		return fmt.Errorf("failed to read response, reason: %v", err)
+	}
+
+	var htsgetURLs htsgetResponse
+	err = json.Unmarshal(body, &htsgetURLs)
+	if err != nil {
+		return fmt.Errorf("error unmarshaling response, reason: %v", err)
+	}
+
+	err = downloadFiles(htsgetURLs, config)
+	if err != nil {
+		return fmt.Errorf("error downloading the files, reason: %v", err)
+	}
+	return nil
+
+}
+
+func downloadFiles(htsgeURLs htsgetResponse, config *helpers.Config) (err error) {
+
+	// Create the file in the outdir or current location
+	var currentPath, pathToUse string
+	currentPath, err = os.Getwd()
+	if err != nil {
+		return fmt.Errorf("failed to get current path, reason: %v", err)
+	}
+	pathToUse = currentPath
+	if outDir != nil && *outDir != "" {
+		pathToUse = *outDir
+	}
+	out, err := os.OpenFile(pathToUse+"/"+*fileName+".c4gh", os.O_CREATE|os.O_WRONLY, 0644)
```

Comment:
When `pathToUse` does not exist, the error message
`reason: open xxxxxx: no such file or directory` is not very informative. 
Better to report the reason by 
`the outdir xxxxxx does not exist`

One can also create the `outdir` if it does not exist. I personally prefer this way.

Reply:
when the filename contains folders, e.g. in the following command. It will exit with error that with `no such file or directory`
```bash
./sda-cli htsget -dataset https://staging-fdp.gdi.nbis.se/dataset/87495812-3201-4099-9478-1c60c9ef8c85/ -key ../starter-kit-storage-and-interfaces/keys/mykey.pub.pem -filename case3/case3_father/Case3_F -reference 22 -htsgethost https://staging-htsget.gdi.nbis.se -config s3cmd.conf   
```

so I think the pathToUse should be handled more properly either `-outdir` is specified or not.

In addition, it will be nice that one can specify the `outfilename` as well. Since now no matter how the user specify value of `-reference`, the outname will be the same.

Reply:
> One can also create the outdir if it does not exist. I personally prefer this way.
I'll add this.

I can see how the error could be improved, but you used a path for a **filename**. The **filename** is primarily used in the request for the file in the form of ` curl http://localhost:8088/reads/DATASET0001/htsnexus_test_NA128781?referenceName=11` and then for the target file of the result.
It makes more sense to me to keep the name consistent and only give the option with `-outdir` to place it elsewhere.
The other way would be something like keeping the filename to use only in the request and have an option like `output` which could be either fullpath name or default to current directory.

Reply:
I'm fine with the strategy you suggested `It makes more sense to me to keep the name consistent and only give the option with -outdir to place it elsewhere.
The other way would be something like keeping the filename to use only in the request and have an option like output which could be either fullpath name or default to current directory.`

Reply:
Hi @pahatz , it's nice that now one can specify the output file name with the flag `-output`. 
However, when one try to retrieve a file with path in the filename, e.g. in the following command

```
$ ./sda-cli htsget -dataset https://staging-fdp.gdi.nbis.se/dataset/87495812-3201-4099-9478-1c60c9ef8c85/ -key c4gh.pub.pem -filename case3/case3_father/Case3_F -reference 22 -htsgethost https://staging-htsget.gdi.nbis.se -config tmp/s3cmd-gdi.conf   
Error: error downloading the files, reason: open /data3/work/BILS/project/project-sda/sda-cli/case3/case3_father/Case3_F.c4gh: no such file or directory
```

It failed because sda-cli is trying to create the file by keeping the original folder structure, but the folder `case3/case3_father/` does not exist locally.

I think it might be better to take one of the following actions 
1. Prompt the user to create the folder `case3/case3_father/` or use the flag `-output` if they want to save the file in the current path in the error message
2. create the folder  `case3/case3_father/` if it does not exist. 

Except for this, the PR looks great now.